tests/suites/test_suite_mpi.function

```diff
@@ -1720,6 +1720,378 @@ exit:
+    mbedtls_mpi_uint *Xa = NULL;
+    mbedtls_mpi_uint *da = NULL;
+
+    mbedtls_mpi_init( &X4 );
+    mbedtls_mpi_init( &X8 );
+    mbedtls_mpi_init( &l );
+    mbedtls_mpi_init( &r );
+
+    TEST_EQUAL( mbedtls_test_read_mpi( &l, input_l ), 0 );
+    TEST_EQUAL( mbedtls_test_read_mpi( &r, input_r ), 0 );
+    TEST_EQUAL( mbedtls_test_read_mpi( &X4, input_X4 ), 0 );
+    TEST_EQUAL( mbedtls_test_read_mpi( &X8, input_X8 ), 0 );
+
+    /* We only need to work with one of (X4, carry4) or (X8, carry8) depending
+     * on sizeof(mbedtls_mpi_uint)
+     */
+    mbedtls_mpi *X = ( sizeof(mbedtls_mpi_uint) == 4 ) ? &X4 : &X8;
+    mbedtls_mpi_uint carry = ( sizeof(mbedtls_mpi_uint) == 4 ) ? carry4 : carry8;
+
+    /* All of the inputs are +ve (or zero) */
+    TEST_EQUAL( l.s, 1 );
+    TEST_EQUAL( r.s, 1 );
+    TEST_EQUAL( X->s, 1 );
+
+    /* Test cases are such that l <= r, so #limbs should be <= */
+    TEST_ASSERT( l.n <= r.n );
+    TEST_ASSERT( X->n <= r.n );
+
+    /* Now let's get arrays of mbedtls_mpi_uints, rather than MPI structures */
+    la = mbedtls_calloc( r.n, sizeof(mbedtls_mpi_uint) );
+    ra = mbedtls_calloc( r.n, sizeof(mbedtls_mpi_uint) );
+    Xa = mbedtls_calloc( r.n, sizeof(mbedtls_mpi_uint) );
+    da = mbedtls_calloc( r.n, sizeof(mbedtls_mpi_uint) );
+
+    TEST_ASSERT( la != NULL );
+    TEST_ASSERT( ra != NULL );
+    TEST_ASSERT( Xa != NULL );
+    TEST_ASSERT( da != NULL );
+
+    /* Populate the arrays. As the mbedtls_mpi_uint[]s in mbedtls_mpis (and as
+     * processed by mbedtls_mpi_core_add_if()) are little endian, we can just
+     * copy what we have as long as MSBs are 0 (which they are from calloc())
+     */
+    memcpy( la, l.p, l.n * sizeof(mbedtls_mpi_uint) );
+    memcpy( ra, r.p, r.n * sizeof(mbedtls_mpi_uint) );
+    memcpy( Xa, X->p, X->n * sizeof(mbedtls_mpi_uint) );
+
+    /* 1a) a += b, cond = 0 => there should be no carry */
+    memcpy( da, la, r.n * sizeof(mbedtls_mpi_uint) );
+    TEST_EQUAL( mbedtls_mpi_core_add_if( da, ra, r.n, 0 ), 0 );
+
+    /* 1b) and a should be unchanged */
+    TEST_EQUAL( memcmp( da, la, r.n * sizeof(mbedtls_mpi_uint) ), 0 );
+
+    /* 2a) a += b, cond = 1 => we should get the correct carry */
+    TEST_EQUAL( mbedtls_mpi_core_add_if( da, ra, r.n, 1 ), carry );
+
+    /* 2b) and a should have the correct result */
+    TEST_EQUAL( memcmp( da, Xa, r.n * sizeof(mbedtls_mpi_uint) ), 0 );
+
+    /* 3a) b += a, cond = 0 => there should be no carry */
+    memcpy( da, ra, r.n * sizeof(mbedtls_mpi_uint) );
+    TEST_EQUAL( mbedtls_mpi_core_add_if( da, la, r.n, 0 ), 0 );
+
+    /* 3b) and b should be unchanged */
+    TEST_EQUAL( memcmp( da, ra, r.n * sizeof(mbedtls_mpi_uint) ), 0 );
+
+    /* 4a) b += a, cond = 1 => we should get the correct carry */
+    TEST_EQUAL( mbedtls_mpi_core_add_if( da, la, r.n, 1 ), carry );
+
+    /* 4b) and b should have the correct result */
+    TEST_EQUAL( memcmp( da, Xa, r.n * sizeof(mbedtls_mpi_uint) ), 0 );
+
```

Comment:
Missing: testing when the input and output are aliased.

Reply:
Done.

I'll note that the test was written when the documentation (incorrectly) stated that A and B must be disjoint :)